### ISSUE TITLE
Large stack usage especially in self test

### ISSUE THREAD

**Issue by c1728p9** [ARMmbed/mbedtls#650]
The self test function ```mbedtls_sha512_self_test``` is using more than 1K of stack, which is causing some platforms to overflow their main stack during testing on mbed-os. The function ```mbedtls_sha512_process``` is using a lot of stack as well so optimizing this would help reduce some stack pressure.

For reference here is the callstack from the mbed-os tls self test, along with stack sizes:
```
Stack size      Frame info
1056            mbedtls_sha512_process() at sha512.c:228
32              mbedtls_sha512_update() at sha512.c:289
0               mbedtls_sha512_update() at sha512.c:337
48              mbedtls_sha512_finish() at sha512.c:338
1328            mbedtls_sha512_self_test() at sha512.c:475
32              mbedtls_sha512_self_test_test_case() at main.cpp:63
48              utest::v1::Harness::run_next_case() at utest_harness.cpp:338
8               utest_us_ticker_run() at utest_shim.cpp:123
80              utest::v1::Harness::run() at utest_harness.cpp:155
24              main() at main.cpp:100
```

**Comment by sg-** [comment]
@sbutcher-arm @yanesca

**Comment by sbutcher-arm** [comment]
Is this an issue for the next mbed OS? What platforms are we impacting?

**Comment by sg-** [comment]
nrf51 and nrf52 have a 2k stack while all other mbed boards have a 4k stack.

**Comment by sbutcher-arm** [comment]
This issue should (hopefully!) be resolved by @c1728p9's fix in pull request #656. That fix should hopefully save enough stack to solve the immediate problem, although in the longer run more stack can probably be saved with more radical changes.
